src/lib/markbind/src/preprocessors/variablePreprocessor.js

```diff
@@ -162,236 +145,187 @@ class VariablePreprocessor {
+      if (isReservedAttribute) {
+        return false;
+      }
+
+      const isExistingAttribute = !!importedFilePageVariables[attr];
+      if (!isExistingAttribute) {
+        logger.warn(`Invalid inline attribute ${attr} imported in ${filePath} from ${resolvedFilePath}\n`);
+        return false;
+      }
+
+      return true;
+    }).forEach((name) => {
+      pageImportedVariables[name] = importedFilePageVariables[name];
     });
-    return importedVariables;
   }
 
   /**
    * Extract page variables from a page.
-   * @param fileName for error printing
+   * These include all locally declared <variable>s and variables <import>ed from other pages.
+   * @param filePath for error printing
    * @param data to extract variables from
-   * @param includedVariables variables from parent include
+   * @param includeVariables from the parent include, if any, used during {@link renderIncludeFile}
    */
-  extractPageVariables(fileName, data, includedVariables) {
-    const fileDir = path.dirname(fileName);
-    const $ = cheerio.load(data);
-
-    const importedVariables = VariablePreprocessor.extractPseudoImportedVariables($);
+  extractPageVariables(filePath, data, includeVariables = {}) {
     const pageVariables = {};
-    const userDefinedVariables = this.getContentParentSiteVariables(fileName);
+    const pageImportedVariables = {};
 
-    const setPageVariableIfNotSetBefore = (variableName, rawVariableValue) => {
-      if (pageVariables[variableName]) {
-        return;
-      }
+    const $ = cheerio.load(data);
 
-      pageVariables[variableName] = njUtil.renderRaw(rawVariableValue, {
-        ...importedVariables,
+    /*
+     This is used to render extracted variables before storing.
+     Hence, variables can be used within other variables, subject to declaration order.
+     */
+    const renderVariable = (contentFilePath, content) => {
+      const previousVariables = {
+        ...pageImportedVariables,
         ...pageVariables,
-        ...userDefinedVariables,
-        ...includedVariables,
-      }, {}, false);
+        ...includeVariables,
+      };
+
+      return this.renderSiteVariables(contentFilePath, content, previousVariables);
     };
 
-    $('variable').each((index, node) => {
-      const variableSource = $(node).attr('from');
-      if (variableSource !== undefined) {
-        try {
-          const variableFilePath = path.resolve(fileDir, variableSource);
-          const jsonData = fs.readFileSync(variableFilePath);
-          const varData = JSON.parse(jsonData);
-          Object.entries(varData).forEach(([varName, varValue]) => {
-            setPageVariableIfNotSetBefore(varName, varValue);
-          });
-        } catch (err) {
-          logger.warn(`Error ${err.message}`);
-        }
+    // NOTE: Selecting both at once is important to respect variable/import declaration order
+    $('variable, import[from]').not('include > import').each((index, elem) => {
```

Comment:
Should this be `include > variable` instead, to avoid recognizing include variables as page variables?